source/hibernated/hql.d

```diff
@@ -0,0 +1,249 @@
+    ParenExpression <- '(' :spaces? Expression :spaces? ')'
+    Unary1Expression <- ( '+' / '-' ) !NumberLit Expression
+    Unary2Expression <- ( NotKw ) :spaces Expression
+
+    SubQueryExpression <- ExistsKw :spaces? :'(' :spaces? SubQuery :spaces? :')'
+                          / Expression :spaces ( InKw / NotInKw ) :spaces :'(' SubQuery :')'
+
+    Identifier      <~ ((!Kw identifier) / (Kw identifier))
+    IdentifierItem  <- Identifier ( :'.' Identifier )*
+    LitItem         <- StringLit / NumberLit / BoolLit / NullLit
+    NamedParamItem  <- ':' identifier
+
+    # See https://learn.microsoft.com/en-us/sql/odbc/reference/appendixes/numeric-literal-syntax?view=sql-server-ver16
+    NumberLit       <- NumExpLit / SignedNumLit
+    SignedNumLit    <~ [-+]? ;UnsignedNumLit
+    UnsignedNumLit  <~ UnsignedInt ( '.' UnsignedInt? )?
+    NumExpLit       <~ SignedNumLit [Ee] SignedInt
+    SignedInt       <- [-+]? ;UnsignedInt
+    UnsignedInt     <- [0-9]+
+
+    # Use the syntactic predicate '!' to fail a match if it starts with "'".
+    StringLit       <- "'" ( "''" / ( ! "'" . ) )* "'"
+
+    BoolLit         <- TrueKw / FalseKw
+    NullLit         <- NullKw
+
+    Kw              <~ AndKw / AsKw / AscKw / AvgKw / BetweenKw / ByKw / CountKw / DeleteKw / DescKw / ExistsKw
+                       / FalseKw / FromKw / InKw / ILikeKw / IsKw / LikeKw / MaxKw / MinKw / NotKw
+                       / OrderKw / OrKw / SelectKw / SumKw / TrueKw / WhereKw
+    AndKw           <~ [Aa][Nn][Dd]
+    AsKw            <~ [Aa][Ss]
+    AscKw           <~ [Aa][Ss][Cc]
+    AvgKw           <~ [Aa][Vv][Gg]
+    BetweenKw       <~ [Bb][Ee][Tt][Ww][Ee][Ee][Nn]
+    ByKw            <~ [Bb][Yy]
+    CountKw         <~ [Cc][Oo][Uu][Nn][Tt]
+    DeleteKw        <~ [Dd][Ee][Ll][Ee][Tt][Ee]
+    DescKw          <~ [Dd][Ee][Ss][Cc]
+    ExistsKw        <~ [Ee][Xx][Ii][Ss][Tt][Ss]
+    FalseKw         <~ [Ff][Aa][Ll][Ss][Ee]
+    FromKw          <~ [Ff][Rr][Oo][Mm]
+    ILikeKw         <~ [Ii][Ll][Ii][Kk][Ee]
+    InKw            <~ [Ii][Nn]
+    IsKw            <~ [Ii][Ss]
+    IsNotKw         <~ [Ii][Ss] :spaces [Nn][Oo][Tt]
+    LikeKw          <~ [Ll][Ii][Kk][Ee]
+    MaxKw           <~ [Mm][Aa][Xx]
+    MinKw           <~ [Mm][Ii][Nn]
+    NotKw           <~ [Nn][Oo][Tt]
+    NotInKw         <~ [Nn][Oo][Tt] :spaces [Ii][Nn]
+    NullKw          <~ [Nn][Uu][Ll][Ll]
+    OrKw            <~ [Oo][Rr]
+    OrderKw         <~ [Oo][Rr][Dd][Ee][Rr]
+    SelectKw        <~ [Ss][Ee][Ll][Ee][Cc][Tt]
+    SimilarKw       <~ [Ss][Ii][Mm][Ii][Ll][Aa][Rr]
+    SumKw           <~ [Ss][Uu][Mm]
+    TrueKw          <~ [Tt][Rr][Uu][Ee]
+    UpdateKw        <~ [Uu][Pp][Dd][Aa][Tt][Ee]
+    WhereKw         <~ [Ww][Hh][Ee][Rr][Ee]
+
+    # End of input, e.g. not any character.
+    identifierChar  <- [a-zA-Z_0-9]
+    eow             <- !identifierChar
+    eoi             <- !.
+`));
+
+/// A sanity check on the HQL select clause.
```

Comment:
@SingingBush So far so good with the parsing, right?

If I continue work to integrate it into the HQL query processing logic in query.d, it should be ok, right?

Reply:
Hi, sorry for the delay in responding. I think if this is merged you will be the primary user that is going to battle test it so I think it's worth keeping as a branch until it's put through it's paces a bit.

I am weary of needing a dependency on _pegged_ by default so I think it worth having hql support either being a configuration that can be added during the build with something like:

```
    "configurations": [
        ...
        {
            "name": "hql",
            "dependencies": {
                "ddbc": "~>0.6.0",
                "pegged": "~>0.4.9"
            }
        }
    ]
```

or having it be it's own sub-project. Then the feature can be something that users choose to opt-in to.

As mentioned before this isn't my repository so having some feed back from buggins or other users would be helpful. Especially as I'm not currently using D for anything critical at the minute.

That said, if integration tests cover the hql feature adequately and everything works as expected then I'm happy to merge it.